sdks/java/io/csv/src/test/java/org/apache/beam/sdk/io/csv/CsvIOWriteTest.java

```diff
@@ -0,0 +1,534 @@
+                .apply(CsvIO.writeRows().to(to).withSchemaFields(Collections.emptyList())));
+
+    assertThrows(
+        "field names should match schema",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of(rows))
+                .setRowSchema(ALL_DATA_TYPES_SCHEMA)
+                .apply(
+                    CsvIO.writeRows()
+                        .to(to)
+                        .withSchemaFields(Arrays.asList("doesnotexist", "string", "anInt"))));
+  }
+
+  @Test
+  public void withCompression() {
+    String to = "with_compression";
+    String regex = String.format("^%s.*$", to);
+    pipeline
+        .apply(Create.of(rows).withRowSchema(ALL_DATA_TYPES_SCHEMA))
+        .apply(
+            CsvIO.writeRows()
+                .to(tmpFolder.getRoot().getAbsolutePath() + "/" + to)
+                .withCompression(Compression.GZIP));
+    pipeline.run().waitUntilFinish();
+    String[] files = filesMatching(regex);
+    for (String name : files) {
+      Path p = Paths.get(tmpFolder.getRoot().getAbsolutePath(), name);
+      File f = new File(p.toString());
+      assertTrue(Compression.GZIP.isCompressed(f.getAbsolutePath()));
+    }
+  }
+
+  @Test
+  public void withInvalidType() {
+    assertThrows(
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of("1,2,3", "4.5.6"))
+                .apply(
+                    CsvIO.<String>write().to(tmpFolder.getRoot().getAbsolutePath() + "/badtype")));
+  }
+
+  @Test
+  public void withNumShards() {
+    String to = "with_num_shards";
+    String regex = String.format("^%s.*$", to);
+    pipeline
+        .apply(Create.of(rows).withRowSchema(ALL_DATA_TYPES_SCHEMA))
+        .apply(
+            CsvIO.writeRows()
+                .to(tmpFolder.getRoot().getAbsolutePath() + "/" + to)
+                .withNumShards(1));
+    pipeline.run().waitUntilFinish();
+    String[] files = filesMatching(regex);
+    assertEquals(1, files.length);
+  }
+
+  @Test
+  public void withPreamble() throws IOException {
+    String preamble =
+        "id=44ebe8ee-dc0f-4c93-8b47-2e627e1eb988\ndate=2022-01-02\ntest=e3bbfc47-9bcb-46f0-8944-829827ee5222";
+    String to = "rows_preamble";
+    String regex = String.format("^%s.*$", to);
+    pipeline
+        .apply(Create.of(rows).withRowSchema(ALL_DATA_TYPES_SCHEMA))
+        .apply(
+            CsvIO.writeRows()
+                .to(tmpFolder.getRoot().getAbsolutePath() + "/" + to)
+                .withPreamble(preamble));
+    pipeline.run().waitUntilFinish();
+    String[] files = filesMatching(regex);
+    assertTrue(
+        "CsvIO.writeRows should write to " + tmpFolder.getRoot().getAbsolutePath() + "/" + to,
+        files.length > 0);
+    String expectedHeader =
+        "aBoolean,aByte,aDouble,aFloat,aLong,aShort,anInt,dateTime,decimal,string";
+    for (String name : files) {
```

Comment:
why would we expect all files to have the same contents?

I think what you want to do is check that the header/preamble are correct and return the rest that is found in each file and then combine all that across all the files and check that it matches expected

here and other places where you use `assertFileContentsMatchInAnyOrder`